### PR TITLE
Introduced PHP 8 support; removed support for PHP 5.6 + 7.0

### PR DESCRIPTION
**Pull request by k00ni**
**Updated 2021-03-10**:

* Introduced **PHP 8** support (mostly test related changes)
* ❗ **Removed support for PHP 5.6 + 7.0**
* Extended **ElementXRef::equals** to allow equality checks in PHP 8+, otherwise tests would fail because `5_0` != `5` (see TODOs below)
* **Separated dev-tooling** (e.g. PHPUnit) from the library by using a dedicated folder named `dev-tools`, which contains a separate `composer.json` file. A detailed explanation why I did that can be found in https://github.com/smalot/pdfparser/pull/383#issuecomment-795322159

**TODOs:**
* [ ] Feedback on the following addition by @smalot: https://github.com/smalot/pdfparser/pull/383/files#diff-d419ba202a4a50b99b1a7004d91d330081176213a0e92f696ec881f3cfefa169R59-R76
* [ ] Create new major version after merging this

### CODE REVIEW

**Comment by k00ni** [comment]
Note to myself:
* [ ] merge in master

**Review by j0k3r — APPROVED** [review]
> I am not sure about the following addition

If that fix the issue, I'm ok with it.

**Comment by k00ni** [comment]
> If that fix the issue, I'm ok with it.

It fixes the test for specific parameters, but I am not sure if it covers it 100%. Can you give me a hint what range of parameter values `equals` has to cover?

**Comment by j0k3r** [comment]
> Can you give me a hint what range of parameter values `equals` has to cover?

I've no idea...

**Comment by k00ni** [comment]
Maybe @Connum?

**Comment by k00ni** [comment]
We should wait with a merge until further information from #104. Its about a memory leak, but some still use PHP 5.6. I would like to sort it out before we cut off PHP 5.6 + 7.0 support.

File: src/Smalot/PdfParser/Element/ElementXRef.php [hunk 1]
```diff
@@ -56,6 +56,24 @@ public function getObject()
...
+         * would fail (= 5_0 and 5 are not equal in PHP 8.0+).
+         */
+        if (
+            true === is_numeric($value)
+            && true === \is_string($this->getContent())
+            && 1 === preg_match('/[0-9]+\_[0-9]+/', $this->getContent(), $matches)
+        ) {
+            return (float) ($this->getContent()) == $value;
+        }
+
```
Comment (k00ni) [inline src/Smalot/PdfParser/Element/ElementXRef.php:76]: @smalot Do you see any issues with that addition?

File: DEVELOPER.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+## Development Tools and Tests
+
+Our test related files are located in `tests` folder.
+Tests are written using PHPUnit.
+
+To install (and update) development tools like PHPUnit or PHP-CS-Fixer run:
+
+> make install-dev-tools
+
+Development toosl are getting installed in `dev-tools/vendor`.
```
Comment (j0k3r) [inline DEVELOPER.md:16]:
```suggestion
Development tools are getting installed in `dev-tools/vendor`.
```

**Review by j0k3r — COMMENTED** [review]
What the benefit of moving dev tools deps in a dedicated folder?

File: DEVELOPER.md (hunk 2 above)
Comment (k00ni) [reply DEVELOPER.md]: Good catch

**Comment by k00ni** [comment]
Semantic Versioning is not clear what happened if a supported environment (in this case PHP 5.6 + 7.0) is dropped. I found a few issues discussing this, like https://github.com/semver/semver/issues/526#issuecomment-519616743.

I would follow the argument and create a new major version (= 1.0.0). 0.18.2 would be the last `0.x` version.

**Comment by smalot** [comment]
Hi @k00ni 
I agree with you to create a new major release "1.x" and to keep old release under the "0.x".
It seems to be the way to reduce any risk of BC Break for current users.

File: src/Smalot/PdfParser/Element/ElementXRef.php (hunk 1 above)
Comment (smalot) [reply src/Smalot/PdfParser/Element/ElementXRef.php:76]: Can you remind me when does this method is called and why we try to validate equality between "5" and "5_0" ?

File: .github/workflows/continuous-integration.yml [hunk 3]
```diff
@@ -107,16 +100,16 @@ jobs:
       - name: "Install dependencies with Composer"
         uses: "ramsey/composer-install@v1"
 
-      - name: "Run PHPUnit (with coverage)"
-        run: "php vendor/bin/simple-phpunit -v --coverage-clover coverage/clover.xml"
+      - name: "Install dev tools"
+        run: "make install-dev-tools"
+
+      - name: "Run PHPUnit"
+        run: "php dev-tools/vendor/bin/phpunit -v --coverage-clover coverage/clover.xml"
```
Comment (smalot) [inline .github/workflows/continuous-integration.yml:107]: Should it be better to use `make run-phpunit ....` instead to hide complexity ?

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+	composer update --working-dir=dev-tools
+
+run-php-cs-fixer:
+	dev-tools/vendor/bin/php-cs-fixer fix
+
+run-phpstan:
+	dev-tools/vendor/bin/phpstan/phpstan/phpstan analyse
+
+run-phpunit:
+	dev-tools/vendor/bin/phpunit
```
Comment (smalot) [inline Makefile:11]: may be allow parameters to be added at the end of command line

**Review by smalot — COMMENTED** [review]
I need to understand why ElementXRef contains this fix

File: src/Smalot/PdfParser/Element/ElementXRef.php (hunk 1 above)
Comment (k00ni) [reply src/Smalot/PdfParser/Element/ElementXRef.php:76]: If you remove this code, [assertion](https://github.com/smalot/pdfparser/blob/php8/tests/Integration/Element/ElementXRefTest.php#L115) in `testEquals` of ElementXRefTest.php will fail.

File: Makefile (hunk 4 above)
Comment (k00ni) [reply Makefile]: Done using optional `$(ARGS)` parameter, see https://github.com/smalot/pdfparser/pull/383/commits/a65d972f0a30ed7fe5e95fdf7e307893099c3fef

If you have a better idea how to do that, please tell me. I am not a Makefile geek.

File: .github/workflows/continuous-integration.yml (hunk 3 above)
Comment (k00ni) [reply .github/workflows/continuous-integration.yml]: Done using optional `$(ARGS)` parameter, see https://github.com/smalot/pdfparser/pull/383/commits/a65d972f0a30ed7fe5e95fdf7e307893099c3fef

If you have a better idea how to do that, please tell me. I am not a Makefile geek.

**Review by smalot — APPROVED** [review]

**Comment by smalot** [comment]
Thanks @k00ni, everything seems to be good for me

**Comment by k00ni** [comment]
@smalot Just to be sure we didn't forget something. What about https://github.com/smalot/pdfparser/pull/383/files/f595ac26308b167c3a02c3a6a77abd653c47b61e#r592171760 (my addition to `ElementXRef.php`)?

**Comment by k00ni** [comment]
We may merge #401 and #402 before this so that older installations can used latest feature.

**Comment by k00ni** [comment]
Because there is no activity/progress in referenced issues I am inclined to merge this PR and prepare a major release.

**Pull request merged**